open_schools_platform/organization_management/circles/services.py

```diff
@@ -56,31 +57,54 @@ def query_handler(query: Query, new_status: str, user: User):
-
         circle_access = user.has_perm('circles.circle_access', query.sender)
-        family_access = user.has_perm('families.family_access', query.recipient)
-
-        if not family_access:
-            if new_status != Query.Status.CANCELED:
-                raise NotAcceptable("Circle can only set canceled status")
-        elif not circle_access:
-            if query.status != Query.Status.SENT:
-                raise NotAcceptable("Сan no longer change the query")
-            if new_status == Query.Status.CANCELED:
-                raise NotAcceptable("User cannot cancel query, he can only decline or accept it")
-
-        query_update(query=query, data={"status": new_status})
-        if query.status == Query.Status.ACCEPTED:
-            if query.body is None:
-                raise MethodNotAllowed("put", detail="Query is corrupted")
-            query.body.circle = query.sender
-            query.body.student_profile = query.additional
-            query.body.save()
-
-        return query
+
+        if type(query.recipient) is Family:
+            family_access = user.has_perm('families.family_access', query.recipient)
+
+            if not family_access:
+                if new_status != Query.Status.CANCELED:
+                    raise NotAcceptable("Circle can only set canceled status")
+            elif not circle_access:
+                if query.status != Query.Status.SENT:
+                    raise NotAcceptable("Сan no longer change the query")
+                if new_status == Query.Status.CANCELED:
+                    raise NotAcceptable("User cannot cancel query, he can only decline or accept it")
+
+            query_update(query=query, data={"status": new_status})
+            if query.status == Query.Status.ACCEPTED:
+                if query.body is None:
+                    raise MethodNotAllowed("put", detail="Query is corrupted")
+                query.body.circle = query.sender
+                query.body.student_profile = query.additional
+                query.body.save()
+            return query
+
+        if type(query.recipient) is TeacherProfile:
+            teacher_profile_access = user.has_perm("teachers.teacher_profile_access", query.recipient)
+
+            if teacher_profile_access and circle_access:
+                pass
+            elif teacher_profile_access:
+                if query.status != Query.Status.SENT:
+                    raise NotAcceptable("Сan no longer change the query")
+                if new_status != Query.Status.CANCELED:
+                    raise NotAcceptable("User can only set canceled status")
+            elif circle_access:
+                if new_status == Query.Status.CANCELED:
+                    raise NotAcceptable("Circle cannot cancel query, it can only decline it")
+                if query.status == Query.Status.CANCELED:
+                    raise NotAcceptable("Сan no longer change the query")
```

Comment:
You need to refactor these lines, because they doesn't work if TeacherProfile is recipient and Circle is sender.

For example, here TeacherProfile can only set canceled status, but this is wrong, because he is the one who received this query. So he need to have ability to accept this query or decline it. 

You can see the above query_handler (for Family), so you can take logic from there as a reference.